File: src/cli/commands/query.ts
import { existsSync } from "node:fs";
import type { Command } from "commander";
import { GraphQueries } from "../../core/graph/queries.js";
import { Registry } from "../../core/registry.js";
import { log } from "../../utils/logger.js";
import { graphDbPath } from "../../utils/paths.js";
import { initDatabase } from "../../core/graph/schema.js";

function openQueries(slug: string): GraphQueries | null {
	const dbPath = graphDbPath(slug);
	if (!existsSync(dbPath)) {
		log.error(`No graph for "${slug}". Run: graphmind build ${slug}`);
		return null;
	}
	const db = initDatabase(dbPath);
	return new GraphQueries(db);
}

function resolveProject(slug: string | undefined, inProject?: string): string | null {
	const registry = new Registry();
	if (inProject) return inProject;
	if (slug) return slug;
	const project = registry.findByPath(process.cwd());
	if (project) return project.slug;
	log.error("Not in a registered project. Specify --in <slug> or register this directory.");
	return null;
}

export function registerQueryCommand(program: Command): void {
	program
		.command("query <symbol>")
		.description("Find a symbol and its connections")
		.option("--in <slug>", "Scope to a specific project")
		.action((symbol: string, opts: { in?: string }) => {
			const slug = resolveProject(undefined, opts.in);
			if (!slug) { process.exitCode = 1; return; }

			const q = openQueries(slug);
			if (!q) { process.exitCode = 1; return; }

			const symbols = q.findSymbol(symbol);
			if (symbols.length === 0) {
				log.dim(`No symbol "${symbol}" found in ${slug}`);
				return;
			}

			for (const s of symbols) {
				console.log(`\n  ${s.kind} ${s.name}  ${s.file}:${s.line_start}`);
				if (s.signature) console.log(`  ${s.signature}`);
			}

			const callers = q.callers(symbol);
			if (callers.length > 0) {
				console.log(`\n  Called by:`);
				for (const c of callers) {
					console.log(`    ${c.name} (${c.file}:${c.line_start})`);
				}
			}

			const callees = q.callees(symbol);
			if (callees.length > 0) {
				console.log(`\n  Calls:`);
				for (const c of callees) {
					console.log(`    ${c.name} (${c.file}:${c.line_start})`);
				}
			}
			console.log();
		});

	program
		.command("fn <symbol>")
		.description("Function call chain + callers")
		.option("--in <slug>", "Scope to project")
		.action((symbol: string, opts: { in?: string }) => {
			const slug = resolveProject(undefined, opts.in);
			if (!slug) { process.exitCode = 1; return; }

			const q = openQueries(slug);
			if (!q) { process.exitCode = 1; return; }

			const symbols = q.findSymbol(symbol);
			if (symbols.length === 0) {
				log.dim(`No symbol "${symbol}" found`);
				return;
			}

			console.log(`\n  ${symbol}`);
			const callers = q.callers(symbol);
			if (callers.length > 0) {
				console.log("  ← called by:");
				for (const c of callers) console.log(`    ${c.name} (${c.file}:${c.line_start})`);
			}
			const callees = q.callees(symbol);
			if (callees.length > 0) {
				console.log("  → calls:");
				for (const c of callees) console.log(`    ${c.name} (${c.file}:${c.line_start})`);
			}
			console.log();
		});

	program
		.command("deps <file>")
		.description("File-level import/export dependency map")
		.option("--in <slug>", "Scope to project")
		.action((file: string, opts: { in?: string }) => {
			const slug = resolveProject(undefined, opts.in);
			if (!slug) { process.exitCode = 1; return; }

			const q = openQueries(slug);
			if (!q) { process.exitCode = 1; return; }

			const deps = q.fileDeps(file);
			const rdeps = q.fileReverseDeps(file);

			if (deps.length > 0) {
				console.log(`\n  ${file} depends on:`);
				for (const d of deps) console.log(`    ${d.file} (${d.kind}: ${d.count})`);
			}
			if (rdeps.length > 0) {
				console.log(`\n  Depended on by:`);
				for (const d of rdeps) console.log(`    ${d.file} (${d.kind}: ${d.count})`);
			}
			if (deps.length === 0 && rdeps.length === 0) {
				log.dim(`No dependencies found for ${file}`);
			}
			console.log();
		});

	program
		.command("impact <file>")
		.description("Transitive reverse dependency trace")
		.option("--in <slug>", "Scope to project")
		.option("--depth <n>", "Max depth", "5")
		.action((file: string, opts: { in?: string; depth: string }) => {
			const slug = resolveProject(undefined, opts.in);
			if (!slug) { process.exitCode = 1; return; }

			const q = openQueries(slug);
			if (!q) { process.exitCode = 1; return; }

			const impacted = q.impact(file, parseInt(opts.depth, 10));
			if (impacted.length === 0) {
				log.dim(`No transitive dependents found for ${file}`);
				return;
			}
			console.log(`\n  Changing ${file} may affect ${impacted.length} file(s):`);
			for (const f of impacted) console.log(`    ${f}`);
			console.log();
		});

	program
		.command("fn-impact <symbol>")
		.description("Blast radius if this function changes")
		.option("--in <slug>", "Scope to project")
		.option("--no-tests", "Exclude test files")
		.action((symbol: string, opts: { in?: string; tests?: boolean }) => {
			const slug = resolveProject(undefined, opts.in);
			if (!slug) { process.exitCode = 1; return; }

			const q = openQueries(slug);
			if (!q) { process.exitCode = 1; return; }

			const symbols = q.findSymbol(symbol);
			if (symbols.length === 0) {
				log.dim(`No symbol "${symbol}" found`);
				return;
			}

			let callers = q.callers(symbol);
			if (opts.tests === false) {
				callers = callers.filter(
					(c) => !c.file.includes(".test.") && !c.file.includes(".spec."),
				);
			}

			console.log(`\n  Blast radius for ${symbol}: ${callers.length} direct caller(s)`);
			for (const c of callers) console.log(`    ${c.name} (${c.file}:${c.line_start})`);
			console.log();
		});

	program
		.command("map [slug]")
		.description("Top N most-connected files")
		.option("-n, --limit <n>", "Number of results", "20")
		.action((slug: string | undefined, opts: { limit: string }) => {
			const resolved = resolveProject(slug);
			if (!resolved) { process.exitCode = 1; return; }

			const q = openQueries(resolved);
			if (!q) { process.exitCode = 1; return; }

			const top = q.topConnected(parseInt(opts.limit, 10));
			console.log(`\n  Top connected files in ${resolved}:`);
			for (const t of top) {
				console.log(`    ${t.connections.toString().padStart(4)} │ ${t.file}`);
			}
			console.log();
		});

	program
		.command("cycles [slug]")
		.description("Detect circular dependencies")
		.action((slug: string | undefined) => {
			const resolved = resolveProject(slug);
			if (!resolved) { process.exitCode = 1; return; }

			const q = openQueries(resolved);
			if (!q) { process.exitCode = 1; return; }

			const cycles = q.detectCycles();
			if (cycles.length === 0) {
				log.success("No circular dependencies detected");
				return;
			}

			log.warn(`${cycles.length} circular dependency pair(s) found:`);
			for (const c of cycles) {
				console.log(`    ${c.from_file} ↔ ${c.to_file}`);
			}
			console.log();
		});
}

File: crates/graphmind-cli/src/commands/query.rs
use graphmind_config::{paths, resolve_project_slug};
use colored::Colorize;
use graphmind_db::queries::GraphQueries;
use graphmind_db::schema::init_database;

fn open_db(slug: &str) -> rusqlite::Connection {
    let db_path = paths::graph_db_path(slug);
    let db_path_str = db_path.to_string_lossy().to_string();
    if !db_path.exists() {
        eprintln!(
            "{} No graph found for {}. Run 'graphmind build' first.",
            "Error:".red().bold(),
            slug
        );
        std::process::exit(1);
    }
    init_database(&db_path_str).unwrap_or_else(|e| {
        eprintln!("{} Failed to open database: {}", "Error:".red().bold(), e);
        std::process::exit(1);
    })
}

pub fn query_symbol(name: &str, slug: Option<&str>) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let symbols = q.find_symbol(name);
    if symbols.is_empty() {
        println!("{} No symbol found: {}", "!".yellow(), name);
        return;
    }

    for s in &symbols {
        println!(
            "{} {} [{}] in {}:{}",
            ">>".cyan().bold(),
            s.name.bold(),
            s.kind.yellow(),
            s.file.dimmed(),
            s.line_start
        );
        if let Some(ref sig) = s.signature {
            if !sig.is_empty() {
                println!("  {}", sig.dimmed());
            }
        }

        // Show callers/callees
        let callers = q.callers(&s.name);
        if !callers.is_empty() {
            println!("  {} ({}):", "Callers".green(), callers.len());
            for c in &callers {
                println!("    {} [{}] {}", c.name, c.edge_kind.dimmed(), c.file.dimmed());
            }
        }
        let callees = q.callees(&s.name);
        if !callees.is_empty() {
            println!("  {} ({}):", "Callees".green(), callees.len());
            for c in &callees {
                println!("    {} [{}] {}", c.name, c.edge_kind.dimmed(), c.file.dimmed());
            }
        }
    }
}

pub fn fn_detail(name: &str, slug: Option<&str>, no_tests: bool) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let symbols = q.find_symbol(name);
    if symbols.is_empty() {
        println!("{} No symbol found: {}", "!".yellow(), name);
        return;
    }

    for s in &symbols {
        if no_tests && (s.file.contains("test") || s.file.contains("spec")) {
            continue;
        }

        println!(
            "{} {} [{}] in {}:{}-{}",
            ">>".cyan().bold(),
            s.name.bold(),
            s.kind.yellow(),
            s.file.dimmed(),
            s.line_start,
            s.line_end
        );
        if let Some(ref sig) = s.signature {
            if !sig.is_empty() {
                println!("  sig: {}", sig);
            }
        }
        if let Some(ref doc) = s.doc {
            if !doc.is_empty() {
                println!("  doc: {}", doc.dimmed());
            }
        }
        if let Some(ref content) = s.content {
            if !content.is_empty() {
                println!("{}", content);
            }
        }
    }
}

pub fn deps(file: &str, slug: Option<&str>) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let deps = q.file_deps(file);
    let rev_deps = q.file_reverse_deps(file);

    if deps.is_empty() && rev_deps.is_empty() {
        println!("{} No dependencies found for {}", "!".yellow(), file);
        return;
    }

    if !deps.is_empty() {
        println!("{} {} depends on:", ">>".cyan().bold(), file.bold());
        for d in &deps {
            println!("  {} [{}] x{}", d.file, d.kind.dimmed(), d.count);
        }
    }

    if !rev_deps.is_empty() {
        println!("\n{} Depended on by:", "<<".cyan().bold());
        for d in &rev_deps {
            println!("  {} [{}] x{}", d.file, d.kind.dimmed(), d.count);
        }
    }
}

pub fn impact(file: &str, slug: Option<&str>, depth: usize) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let impacted = q.impact(file, depth);
    if impacted.is_empty() {
        println!("{} No impact found for {}", "!".yellow(), file);
        return;
    }

    println!(
        "{} {} files impacted by changes to {}:",
        ">>".cyan().bold(),
        impacted.len().to_string().green(),
        file.bold()
    );
    for f in &impacted {
        println!("  {f}");
    }
}

pub fn fn_impact(name: &str, slug: Option<&str>, depth: usize) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let symbols = q.find_symbol(name);
    if symbols.is_empty() {
        println!("{} No symbol found: {}", "!".yellow(), name);
        return;
    }

    let mut all_impacted = std::collections::HashSet::new();
    for s in &symbols {
        let impacted = q.impact(&s.file, depth);
        for f in impacted {
            all_impacted.insert(f);
        }
    }

    if all_impacted.is_empty() {
        println!("{} No impact found for symbol {}", "!".yellow(), name);
        return;
    }

    println!(
        "{} {} files impacted by changes to symbol {}:",
        ">>".cyan().bold(),
        all_impacted.len().to_string().green(),
        name.bold()
    );
    for f in &all_impacted {
        println!("  {f}");
    }
}

pub fn map(slug: Option<&str>) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let stats = q.stats();
    let langs = q.language_breakdown();
    let top = q.top_connected(20);

    println!("{} Graph map for {}", ">>".cyan().bold(), slug.cyan());
    println!(
        "  {} symbols | {} edges | {} files",
        stats.symbols.to_string().green(),
        stats.edges.to_string().green(),
        stats.files.to_string().green()
    );

    if !langs.is_empty() {
        println!("\n  {}:", "Languages".bold());
        for l in &langs {
            println!("    {} ({})", l.language, l.count);
        }
    }

    if !top.is_empty() {
        println!("\n  {}:", "Top connected files".bold());
        for t in &top {
            println!("    {} ({})", t.file, t.connections);
        }
    }
}

pub fn cycles(slug: Option<&str>) {
    let slug = require_slug(slug);
    let db = open_db(&slug);
    let q = GraphQueries::new(&db);

    let cycles = q.detect_cycles();
    if cycles.is_empty() {
        println!("{} No cycles detected", "OK".green().bold());
        return;
    }

    println!(
        "{} {} circular dependencies detected:",
        "!".yellow().bold(),
        cycles.len().to_string().red()
    );
    for c in &cycles {
        println!("  {} <-> {}", c.from_file.red(), c.to_file.red());
    }
}

fn require_slug(slug: Option<&str>) -> String {
    match resolve_project_slug(&[slug]) {
        Some(s) => s,
        None => {
            eprintln!(
                "{} No project specified and none could be resolved",
                "Error:".red().bold()
            );
            std::process::exit(1);
        }
    }
}
